{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting max number of codons from all the samples\n",
    "\n",
    "1. dhfr\n",
    "2. mdr1\n",
    "3. dhps\n",
    "4. crt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "largest for PfDHFR 609\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfDHFR/Category_results/avg_cat_cov/'\n",
    "\n",
    "list1 = []\n",
    "for filename in os.listdir(cov_directory): ## take into accoiunt the samples for which i dont have images, give them -2\n",
    "    if filename != \".DS_Store\":\n",
    "        filepath = os.path.join(cov_directory, filename)\n",
    "        sample_num = filename[4:10].split(\"P\")[0]\n",
    "        with open(filepath) as f:\n",
    "            lines = f.readlines()\n",
    "            for i in range(len(lines)):\n",
    "                line = lines[i].split(',')\n",
    "                num_of_codons = len(line) - 1\n",
    "                list1.append(num_of_codons)\n",
    "#                 print(filename, num_of_codons)\n",
    "\n",
    "largest_number = max(list1)\n",
    "print(\"largest for PfDHFR\", largest_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "largest for mdr1 1404\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfMDR1/Category_results/avg_cat_cov/'\n",
    "\n",
    "list1 = []\n",
    "for filename in os.listdir(cov_directory): ## take into accoiunt the samples for which i dont have images, give them -2\n",
    "    if filename != \".DS_Store\":\n",
    "        filepath = os.path.join(cov_directory, filename)\n",
    "        sample_num = filename[4:10].split(\"P\")[0]\n",
    "        with open(filepath) as f:\n",
    "            lines = f.readlines()\n",
    "            for i in range(len(lines)):\n",
    "                line = lines[i].split(',')\n",
    "                num_of_codons = len(line) - 1\n",
    "                list1.append(num_of_codons)\n",
    "#                 print(filename, num_of_codons)\n",
    "\n",
    "largest_number = max(list1)\n",
    "print(\"largest for mdr1\", largest_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "largest for dhps 805\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfDHPS/Category_results/avg_cat_cov/'\n",
    "\n",
    "list1 = []\n",
    "for filename in os.listdir(cov_directory): ## take into accoiunt the samples for which i dont have images, give them -2\n",
    "    if filename != \".DS_Store\":\n",
    "        filepath = os.path.join(cov_directory, filename)\n",
    "        sample_num = filename[4:10].split(\"P\")[0]\n",
    "        with open(filepath) as f:\n",
    "            lines = f.readlines()\n",
    "            for i in range(len(lines)):\n",
    "                line = lines[i].split(',')\n",
    "                num_of_codons = len(line) - 1\n",
    "                list1.append(num_of_codons)\n",
    "#                 print(filename, num_of_codons)\n",
    "\n",
    "largest_number = max(list1)\n",
    "print(\"largest for dhps\", largest_number)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "largest for crt 657\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfCRT/Category_results/avg_cat_cov/'\n",
    "\n",
    "list1 = []\n",
    "for filename in os.listdir(cov_directory): ## take into accoiunt the samples for which i dont have images, give them -2\n",
    "    if filename != \".DS_Store\":\n",
    "        filepath = os.path.join(cov_directory, filename)\n",
    "        sample_num = filename[4:10].split(\"P\")[0]\n",
    "        with open(filepath) as f:\n",
    "            lines = f.readlines()\n",
    "            for i in range(len(lines)):\n",
    "                line = lines[i].split(',')\n",
    "                num_of_codons = len(line) - 1\n",
    "                list1.append(num_of_codons)\n",
    "#                 print(filename, num_of_codons)\n",
    "\n",
    "largest_number = max(list1)\n",
    "print(\"largest for crt\", largest_number)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Categorizing samples based on "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                SPECIES_CT    pool  dhfr  mdr1  crt  dhps\n",
      "Sequences_name                                           \n",
      "9540                 20.99   pool1   1.0   1.0  1.0   0.0\n",
      "9593                 22.01   pool1   1.0   1.0  1.0   1.0\n",
      "9955                 22.02   pool1   1.0   1.0  1.0   0.0\n",
      "9696                 22.06   pool1   1.0   1.0  1.0   0.0\n",
      "9832                 22.30   pool1   1.0   1.0  1.0   0.0\n",
      "10026                22.40   pool1   1.0   1.0  1.0   0.0\n",
      "10076                22.40   pool1   1.0   1.0  1.0   0.0\n",
      "9953                 22.50   pool1   1.0   1.0  1.0   0.0\n",
      "9867                 22.55   pool1   1.0   1.0  1.0   1.0\n",
      "9647                 22.60   pool1   1.0   1.0  1.0   0.0\n",
      "9750                 22.95   pool2   1.0   1.0  1.0   0.0\n",
      "9934                 22.95   pool2   1.0   1.0  1.0   0.0\n",
      "10007                23.27   pool2   1.0   1.0  1.0   0.0\n",
      "10018                23.90   pool2   1.0   1.0  1.0   1.0\n",
      "10118                23.90   pool2   1.0   1.0  1.0   0.0\n",
      "10044                24.42   pool2   1.0   1.0  1.0   0.0\n",
      "9918                 24.50   pool2   1.0   1.0  1.0   1.0\n",
      "9816                 24.60   pool2   1.0   1.0  1.0   0.0\n",
      "9654                 24.70   pool2   1.0   1.0  1.0   1.0\n",
      "10022                24.75   pool2   1.0   1.0  1.0   0.0\n",
      "9688                 25.20   pool3  -3.0  -3.0 -3.0  -3.0\n",
      "9743                 25.30   pool3   1.0   1.0  1.0   0.0\n",
      "9706                 25.37   pool3   1.0   1.0  1.0   0.0\n",
      "9572                 25.50   pool3   1.0   1.0  1.0   0.0\n",
      "9556                 25.60   pool3   1.0   0.0 -1.0  -2.0\n",
      "9930                 25.70   pool3   1.0   1.0  1.0   0.0\n",
      "9546                 25.80   pool3   1.0   1.0  1.0   0.0\n",
      "9791                 25.90   pool3   1.0   1.0  1.0   0.0\n",
      "9873                 26.00   pool3   1.0   1.0  1.0   0.0\n",
      "10079                26.60   pool3   1.0   1.0  1.0   0.0\n",
      "9646                 27.00   pool4   1.0   1.0  1.0   0.0\n",
      "9916                 27.10   pool4   1.0   1.0  1.0   0.0\n",
      "10000                27.20   pool4   1.0   1.0  1.0   0.0\n",
      "9869                 27.70   pool4   1.0   1.0  1.0   0.0\n",
      "9748                 27.90   pool4   1.0   1.0  1.0   0.0\n",
      "9799                 28.10   pool4   1.0   1.0  1.0   0.0\n",
      "9637                 28.30   pool4   1.0   1.0  1.0   0.0\n",
      "9808                 28.38   pool4   1.0   1.0  1.0   0.0\n",
      "9984                 28.60   pool4   1.0   1.0  1.0   0.0\n",
      "10163                28.90   pool4   1.0   1.0  1.0   0.0\n",
      "9785                 29.10   pool5   1.0   1.0  1.0   0.0\n",
      "9667                 29.30   pool5   1.0   1.0  1.0   0.0\n",
      "9625                 29.40   pool5   1.0   1.0  1.0   0.0\n",
      "9844                 29.50   pool5   1.0   1.0  1.0   0.0\n",
      "9805                 29.70   pool5   1.0   1.0  1.0   0.0\n",
      "9545                 29.88   pool5   1.0   1.0  1.0   0.0\n",
      "9919                 29.90   pool5   1.0   1.0  0.0  -1.0\n",
      "9943                 30.24   pool5   1.0   1.0  1.0   0.0\n",
      "9838                 30.40   pool5   1.0   1.0  1.0   0.0\n",
      "9780                 30.90   pool5   1.0   1.0  1.0   0.0\n",
      "9975                 31.10   pool6   1.0   1.0  0.0  -1.0\n",
      "9579                 31.20   pool6   1.0   1.0  1.0   0.0\n",
      "9924                 31.20   pool6   1.0   1.0  1.0   0.0\n",
      "9978                 31.40   pool6   1.0   1.0  1.0   0.0\n",
      "9859                 31.50   pool6   1.0   1.0  1.0   0.0\n",
      "9866                 31.60   pool6   1.0   1.0  1.0   0.0\n",
      "10104                32.17   pool6   1.0   1.0 -1.0  -2.0\n",
      "10102                32.50   pool6   1.0   1.0  0.0  -1.0\n",
      "9950                 32.80   pool6   1.0   1.0  1.0   0.0\n",
      "9994                 32.90   pool6   1.0   1.0  1.0   0.0\n",
      "9610                 33.20   pool7   1.0   1.0 -2.0  -2.0\n",
      "9769                 33.31   pool7   1.0   1.0  1.0   0.0\n",
      "9575                 33.50   pool7   1.0   1.0  1.0   0.0\n",
      "9973                 33.80   pool7   1.0   1.0  1.0   0.0\n",
      "9794                 33.90   pool7   1.0   1.0 -2.0  -2.0\n",
      "9599                 34.40   pool7   1.0   1.0  1.0   0.0\n",
      "9656                 34.60   pool7   1.0  -3.0 -2.0  -2.0\n",
      "9754                 34.70   pool7   1.0  -3.0 -2.0  -2.0\n",
      "9819                 34.80   pool7   1.0   0.0  0.0  -1.0\n",
      "9903                 34.90   pool7   1.0  -1.0 -3.0  -3.0\n",
      "9897                 35.00   pool8   1.0   0.0 -2.0  -2.0\n",
      "9683                 35.20   pool8  -2.0  -3.0 -3.0  -3.0\n",
      "9672                 35.50   pool8  -1.0  -2.0 -3.0  -3.0\n",
      "9596                 35.70   pool8   0.0   1.0  1.0  -1.0\n",
      "9779                 35.84   pool8  -3.0  -3.0 -3.0  -3.0\n",
      "9855                 35.90   pool8   1.0   0.0 -1.0  -2.0\n",
      "9944                 36.20   pool8   1.0   0.0  1.0   0.0\n",
      "9788                 36.50   pool8   1.0  -1.0  1.0   0.0\n",
      "9621                 36.90   pool8   0.0  -1.0 -2.0  -2.0\n",
      "9884                 36.90   pool8   1.0  -2.0 -1.0  -2.0\n",
      "10121                37.00   pool9   1.0  -2.0  1.0   0.0\n",
      "9913                 37.10   pool9  -1.0   1.0  0.0  -1.0\n",
      "10100                37.20   pool9   1.0   1.0 -1.0  -2.0\n",
      "10040                37.30   pool9   1.0   1.0 -1.0  -2.0\n",
      "9633                 37.50   pool9  -3.0  -1.0  1.0   0.0\n",
      "10158                37.70   pool9   1.0   1.0 -2.0  -2.0\n",
      "9817                 37.80   pool9   1.0   1.0 -2.0  -2.0\n",
      "10122                38.50   pool9  -3.0   1.0 -1.0  -2.0\n",
      "9560                 38.60   pool9   0.0   1.0 -3.0  -3.0\n",
      "9910                 38.90   pool9  -3.0  -3.0 -3.0  -3.0\n",
      "10133                39.50  pool10   1.0  -3.0 -3.0  -3.0\n",
      "10039                39.69  pool10  -2.0  -3.0 -3.0  -3.0\n",
      "9999                 39.83  pool10   1.0  -2.0 -2.0  -2.0\n",
      "9725                 39.92  pool10  -3.0  -3.0 -3.0  -3.0\n",
      "9568                 40.00  pool10  -3.0  -3.0 -3.0  -3.0\n",
      "9810                 40.10  pool10   0.0  -3.0 -3.0  -3.0\n",
      "10222                40.10  pool10  -3.0  -3.0 -3.0  -3.0\n",
      "9569                 40.20  pool10   1.0  -3.0 -3.0  -3.0\n",
      "10147                40.30  pool10  -2.0  -3.0 -3.0  -3.0\n",
      "9567                 40.40  pool10  -3.0  -3.0 -3.0  -3.0\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import pandas as pd\n",
    "# pd.set_option(\"display.max_rows\", None, \"display.max_columns\", None)\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Categorizing Haiti individual samples based on their coverage graphs and linking them to their CT values\n",
    "(~parasitemia levels).\n",
    "\n",
    "Total samples for Haiti study = 100\n",
    "\n",
    "Separating samples based on quality.\n",
    "\n",
    "Key:\n",
    "\n",
    "-3 ==> No data available. No coverage calculation was done.\n",
    "-2 ==> Negligible data available. Sequencing error.\n",
    "-1 ==> Less number of codons covered for respective genes.\n",
    " 0 ==> Mediocre number of codons covered for respective genes.\n",
    " 1 ==> Maximum or near to maximum number of codons covered for respective genes.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "excel_path = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/Haiti_PetValues.xlsx'\n",
    "\n",
    "### Getting PET values data from tyhe excel file. Selecting just the sequences name.\n",
    "\n",
    "df1 = pd.read_excel(excel_path, sheet_name = \"CovValidation\")\n",
    "seq_name_list = list(df1.Sequences_name)\n",
    "\n",
    "### Replacing Nan values to zero so that data frame can be filled with key values.\n",
    "\n",
    "df2 = pd.read_excel(excel_path, sheet_name = \"CovValidation\", index_col=0)\n",
    "df2.fillna(0, inplace=True)\n",
    "\n",
    "### Getting samples for PfDHFR coverage results prepared to categorize the codon positions\n",
    "\n",
    "def dhfr_values():\n",
    "\n",
    "    filepath_list = []\n",
    "    sample_list = []\n",
    "    cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfDHFR/Category_results/avg_cat_cov/'\n",
    "    for filename in os.listdir(cov_directory):\n",
    "        if filename != \".DS_Store\" and filename != \"test\":\n",
    "            filepath = os.path.join(cov_directory, filename)\n",
    "            filepath_list.append(filepath)\n",
    "            sample_num = int(filename[4:10].split(\"P\")[0]) #Getting sample number/id\n",
    "            sample_list.append(sample_num)\n",
    "\n",
    "    \"\"\"\n",
    "    Total samples for Haiti study = 100\n",
    "    Samples generated from codon coverage calculation for PfDHFR = 91\n",
    "    Max number of codons for PfDHFR = 609\n",
    "\n",
    "    Key for catrgorizing:\n",
    "\n",
    "    -3 = No sample data\n",
    "    -2 = num of codons < 100\n",
    "    -1 = 100=< num of codons < 500\n",
    "     0 = 500=< num of codons < 600\n",
    "     1 = num of codons >= 600\n",
    "    \"\"\"\n",
    "\n",
    "    no_data_samples = []\n",
    "    data_samples = []\n",
    "    for seq_name in seq_name_list:\n",
    "        if seq_name not in sample_list:\n",
    "            no_data_samples.append(seq_name)\n",
    "            df2.at[seq_name, 'dhfr'] = -3 #No data for these samples. No images, nothing.\n",
    "        if seq_name in sample_list:\n",
    "            for sample_name in sample_list:\n",
    "                if seq_name == sample_name:\n",
    "                    data_samples.append(seq_name)\n",
    "\n",
    "\n",
    "    for seq_id in data_samples:\n",
    "        for file in filepath_list:\n",
    "            file_id = int(file[71:77].split(\"P\")[0])\n",
    "            if seq_id == file_id:\n",
    "                with open(file) as f:\n",
    "    #                 print(\"filename\", file)\n",
    "                    lines = f.readlines()\n",
    "                    for i in range(len(lines)):\n",
    "                        line = lines[i].split(',')\n",
    "                        num_of_codons = len(line) - 1\n",
    "    #                     print(\"num of codons\", num_of_codons)\n",
    "                        if num_of_codons >= 600: ## these values will change for each gene\n",
    "                            df2.at[seq_id, 'dhfr'] = 1\n",
    "                        elif num_of_codons >= 500 and num_of_codons < 600:## these values will change for each gene\n",
    "                            df2.at[seq_id, 'dhfr'] = 0\n",
    "                        elif num_of_codons >= 100 and num_of_codons < 500:## these values will change for each gene\n",
    "                            df2.at[seq_id, 'dhfr'] = -1\n",
    "                        elif num_of_codons < 100:## these values will change for each gene\n",
    "                            df2.at[seq_id, 'dhfr'] = -2\n",
    "    return df2\n",
    "\n",
    "dhfr_df = dhfr_values()\n",
    "\n",
    "\n",
    "def mdr1_values(input_dhfr_df):\n",
    "    \n",
    "    filepath_list = []\n",
    "    sample_list = []\n",
    "    cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfMDR1/Category_results/avg_cat_cov/'\n",
    "    for filename in os.listdir(cov_directory):\n",
    "        if filename != \".DS_Store\": #and filename != \"test\":\n",
    "            filepath = os.path.join(cov_directory, filename)\n",
    "            filepath_list.append(filepath)\n",
    "            sample_num = int(filename[4:10].split(\"P\")[0]) #Getting sample number/id\n",
    "            sample_list.append(sample_num)\n",
    "\n",
    "    \"\"\"\n",
    "    Total samples for Haiti study = 100\n",
    "    Samples generated from codon coverage calculation for PfMDR1 = 85\n",
    "    Max number of codons for PfMDR1 = 1404\n",
    "\n",
    "    Key for catergorizing:\n",
    "\n",
    "    -3 = No sample data\n",
    "    -2 = num of codons < 140\n",
    "    -1 = 140=< num of codons < 1050\n",
    "     0 = 1050=< num of codons < 1330\n",
    "     1 = num of codons >= 1330\n",
    "    \"\"\"\n",
    "\n",
    "    no_data_samples = []\n",
    "    data_samples = []\n",
    "    for seq_name in seq_name_list:\n",
    "        if seq_name not in sample_list:\n",
    "            no_data_samples.append(seq_name)\n",
    "            df2.at[seq_name, 'mdr1'] = -3 #No data for these samples. No images, nothing.\n",
    "        if seq_name in sample_list:\n",
    "            for sample_name in sample_list:\n",
    "                if seq_name == sample_name:\n",
    "                    data_samples.append(seq_name)\n",
    "\n",
    "\n",
    "    for seq_id in data_samples:\n",
    "        for file in filepath_list:\n",
    "            file_id = int(file[71:77].split(\"P\")[0])\n",
    "            if seq_id == file_id:\n",
    "                with open(file) as f:\n",
    "    #                 print(\"filename\", file)\n",
    "                    lines = f.readlines()\n",
    "                    for i in range(len(lines)):\n",
    "                        line = lines[i].split(',')\n",
    "                        num_of_codons = len(line) - 1\n",
    "    #                     print(\"num of codons\", num_of_codons)\n",
    "                        if num_of_codons >= 1330: ## these values will change for each gene\n",
    "                            input_dhfr_df.at[seq_id, 'mdr1'] = 1\n",
    "                        elif num_of_codons >= 1050 and num_of_codons < 1330:## these values will change for each gene\n",
    "                            input_dhfr_df.at[seq_id, 'mdr1'] = 0\n",
    "                        elif num_of_codons >= 140 and num_of_codons < 1050:## these values will change for each gene\n",
    "                            input_dhfr_df.at[seq_id, 'mdr1'] = -1\n",
    "                        elif num_of_codons < 140:## these values will change for each gene\n",
    "                            input_dhfr_df.at[seq_id, 'mdr1'] = -2\n",
    "\n",
    "    return input_dhfr_df\n",
    "\n",
    "mdr1_df = mdr1_values(dhfr_df)\n",
    "\n",
    "\n",
    "def dhps_values(input_mdr1_df):\n",
    "\n",
    "    filepath_list = []\n",
    "    sample_list = []\n",
    "    cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfDHPS/Category_results/avg_cat_cov/'\n",
    "    for filename in os.listdir(cov_directory):\n",
    "        if filename != \".DS_Store\": #and filename != \"test\":\n",
    "            filepath = os.path.join(cov_directory, filename)\n",
    "            filepath_list.append(filepath)\n",
    "            sample_num = int(filename[4:10].split(\"P\")[0]) #Getting sample number/id\n",
    "            sample_list.append(sample_num)\n",
    "\n",
    "    \"\"\"\n",
    "    Total samples for Haiti study = 100\n",
    "    Samples generated from codon coverage calculation for PfDHPS = 84\n",
    "    Max number of codons for PfDHPS = 805\n",
    "\n",
    "    Key for catrgorizing:\n",
    "\n",
    "    -3 = No sample data\n",
    "    -2 = num of codons < 300\n",
    "    -1 = 300=< num of codons < 540\n",
    "     0 = 540=< num of codons < 750\n",
    "     1 = num of codons >= 750\n",
    "    \"\"\"\n",
    "\n",
    "    no_data_samples = []\n",
    "    data_samples = []\n",
    "    for seq_name in seq_name_list:\n",
    "        if seq_name not in sample_list:\n",
    "            no_data_samples.append(seq_name)\n",
    "            df2.at[seq_name, 'dhps'] = -3 #No data for these samples. No images, nothing.\n",
    "        if seq_name in sample_list:\n",
    "            for sample_name in sample_list:\n",
    "                if seq_name == sample_name:\n",
    "                    data_samples.append(seq_name)\n",
    "\n",
    "\n",
    "    for seq_id in data_samples:\n",
    "        for file in filepath_list:\n",
    "            file_id = int(file[71:77].split(\"P\")[0])\n",
    "            if seq_id == file_id:\n",
    "                with open(file) as f:\n",
    "    #                 print(\"filename\", file)\n",
    "                    lines = f.readlines()\n",
    "                    for i in range(len(lines)):\n",
    "                        line = lines[i].split(',')\n",
    "                        num_of_codons = len(line) - 1\n",
    "    #                     print(\"num of codons\", num_of_codons)\n",
    "                        if num_of_codons >= 750: ## these values will change for each gene\n",
    "                            input_mdr1_df.at[seq_id, 'dhps'] = 1\n",
    "                        elif num_of_codons >= 540 and num_of_codons < 750:## these values will change for each gene\n",
    "                            input_mdr1_df.at[seq_id, 'dhps'] = 0\n",
    "                        elif num_of_codons >= 300 and num_of_codons < 540:## these values will change for each gene\n",
    "                            input_mdr1_df.at[seq_id, 'dhps'] = -1\n",
    "                        elif num_of_codons < 300:## these values will change for each gene\n",
    "                            input_mdr1_df.at[seq_id, 'dhps'] = -2\n",
    "                            \n",
    "    return input_mdr1_df\n",
    "\n",
    "dhps_df = dhps_values(mdr1_df)\n",
    "\n",
    "def crt_values(input_dhps_df):\n",
    "    filepath_list = []\n",
    "    sample_list = []\n",
    "    cov_directory = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/PfDHPS/Category_results/avg_cat_cov/'\n",
    "    for filename in os.listdir(cov_directory):\n",
    "        if filename != \".DS_Store\": #and filename != \"test\":\n",
    "            filepath = os.path.join(cov_directory, filename)\n",
    "            filepath_list.append(filepath)\n",
    "            sample_num = int(filename[4:10].split(\"P\")[0]) #Getting sample number/id\n",
    "            sample_list.append(sample_num)\n",
    "\n",
    "    \"\"\"\n",
    "    Total samples for Haiti study = 100\n",
    "    Samples generated from codon coverage calculation for PfCRT = 85\n",
    "    Max number of codons for PfCRT = 657\n",
    "    NOTE: There is only one sample with max number of codons. Else it is near 480 codons.\n",
    "\n",
    "    Key for catrgorizing:\n",
    "\n",
    "    -3 = No sample data\n",
    "    -2 = num of codons < 180\n",
    "    -1 = 180=< num of codons < 330\n",
    "     0 = 330=< num of codons < 480\n",
    "     1 = num of codons >= 480\n",
    "    \"\"\"\n",
    "\n",
    "    no_data_samples = []\n",
    "    data_samples = []\n",
    "    for seq_name in seq_name_list:\n",
    "        if seq_name not in sample_list:\n",
    "            no_data_samples.append(seq_name)\n",
    "            df2.at[seq_name, 'crt'] = -3 #No data for these samples. No images, nothing.\n",
    "        if seq_name in sample_list:\n",
    "            for sample_name in sample_list:\n",
    "                if seq_name == sample_name:\n",
    "                    data_samples.append(seq_name)\n",
    "\n",
    "\n",
    "    for seq_id in data_samples:\n",
    "        for file in filepath_list:\n",
    "            file_id = int(file[71:77].split(\"P\")[0])\n",
    "            if seq_id == file_id:\n",
    "                with open(file) as f:\n",
    "    #                 print(\"filename\", file)\n",
    "                    lines = f.readlines()\n",
    "                    for i in range(len(lines)):\n",
    "                        line = lines[i].split(',')\n",
    "                        num_of_codons = len(line) - 1\n",
    "    #                     print(\"num of codons\", num_of_codons)\n",
    "                        if num_of_codons >= 480: ## these values will change for each gene\n",
    "                            input_dhps_df.at[seq_id, 'crt'] = 1\n",
    "                        elif num_of_codons >= 330 and num_of_codons < 480:## these values will change for each gene\n",
    "                            input_dhps_df.at[seq_id, 'crt'] = 0\n",
    "                        elif num_of_codons >= 180 and num_of_codons < 330:## these values will change for each gene\n",
    "                            input_dhps_df.at[seq_id, 'crt'] = -1\n",
    "                        elif num_of_codons < 180:## these values will change for each gene\n",
    "                            input_dhps_df.at[seq_id, 'crt'] = -2\n",
    "    return input_dhps_df\n",
    "\n",
    "crt_df = crt_values(dhps_df)\n",
    "print(crt_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = r'/Users/mansi/Desktop/Haiti/Individual/Codon_coverage/Haiti_PetValues.xlsx'\n",
    "\n",
    "with pd.ExcelWriter(path, mode='a') as writer:\n",
    "    crt_df.to_excel(writer, sheet_name='CovValidation', verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
